File: 백준/Gold/2671. 잠수함식별/잠수함식별.py
import sys
input = sys.stdin.readline

#       0   1
# S0:   1   2
# S1:   -   0
# S2:   3   -
# S3:   4   -
# S4:   4   5
# S5:   1   6
# S6:   7   6
# S7:   4   0

def main():
    re = [(1, 2), (-1, 0), (3, -1), (4, -1), (4, 5), (1, 6), (7, 6), (4, 0)]
    s = input().rstrip()
    state = 0
    for c in s:
        if state == -1:
            break
        i = int(c)
        state = re[state][i]
    if state == 0 or state == 5 or state == 6:
        print("SUBMARINE")
    else:
        print("NOISE")


if __name__ == "__main__":
    main()

File: 백준/Gold/2671. 잠수함식별/잠수함식별.cc
#include <bits/stdc++.h>
using namespace std;

#define MAX_STATE_NUM 10
#define MAX_STRING_SIZE 150

#define START_STATE 0
#define SINK_STATE (MAX_STATE_NUM-1)


string s;
int state = START_STATE;
int transition_table[MAX_STATE_NUM][2] = {
	{5,1},
	{2, SINK_STATE},
	{3, SINK_STATE},
	{3, 4},
	{5, 7},
	{SINK_STATE, 6},
	{5, 1},
	{8, 7},
	{3, 6},
	{SINK_STATE, SINK_STATE}
};

int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0); cout.tie(0);
	cin >> s;

	for (int i = 0; s[i] != '\0'; i++) {
		state = transition_table[state][s[i] - '0'];
	}

	if (state == 4 || state == 6 || state == 7) {
		puts("SUBMARINE");
	}
	else {
		puts("NOISE");
	}
}